### ISSUE TITLE
FP32 `jax.random.categorical` inconsistent with the FP64 version

### ISSUE THREAD

**Issue by zdcao121** [jax-ml/jax#25749]
### Description

The results of the following code are not consistent while switching from `fp32` to `fp64`, which is controlled by the setting `jax.config.update("jax_enable_x64", True)`.

```
import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp

logits = jnp.array([-1.03401458e+00, -2.89790678e+00, -1.78951931e+00,
                    -2.94101143e+00, -3.74539852e+00, -3.86494231e+00, -4.60886765e+00,
                    -3.80349803e+00, -1.22899675e+00])

key = jax.random.PRNGKey(42)
temperature = 1.0
w = jax.random.categorical(key, logits/temperature)

print(jnp.result_type(w), w)
```  

When using `fp32`, the result is `int32 0`, whereas in `fp64`, it is `int64 7`. It seems the `jax.random.categorical` is not stable and can cause huge difference between `fp32` and `fp64`.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.4
python: 3.9.18 | packaged by conda-forge | (main, Dec 23 2023, 16:33:10)  [GCC 12.3.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='node002', release='3.10.0-1062.12.1.el7.x86_64', version='#1 SMP Tue Feb 4 23:02:59 UTC 2020', machine='x86_64')


$ nvidia-smi
Tue Jan  7 17:44:25 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-PCIE-40GB          Off |   00000000:16:00.0 Off |                  Off |
| N/A   28C    P0             37W /  250W |     425MiB /  40960MiB |      3%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A     64664      C   /data/zdcao/envs/torch/bin/python3.9          416MiB |
+-----------------------------------------------------------------------------------------+

```

**Comment by jakevdp** [comment]
Hi - thanks for the report! I think this is probably working as expected. When you change to a different floating point representation, you'll get different floating point results, which in this case are mapped to different categories. I'm not sure there's any way we could change this.

**Comment by zdcao121** [comment]
Thanks for your quick reply!

**Comment by zdcao121** [comment]
btw, I'm curious about whether it will cause the degradation in the inference of LLM because the sampling is heavily based on the `jax.random.categorical`.

**Comment by jakevdp** [comment]
I can't think of any mechanism by which this would degrade LLM inference – is there something in particular you have in mind?

**Comment by zdcao121** [comment]
Switching between different precisions can lead to varying behaviors in `jax.random.categorical`, even with the same logits, which may produce significantly different outputs in LLM sampling.

https://github.com/xai-org/grok-1/blob/main/runners.py#L118

**Comment by jakevdp** [comment]
Sure, I agree that you'll get different results depending on whether you're computing in 32-bit or 64-bit. But I don't think I'd classify that as "degradation". It's just a different result when you're doing a different computation.

**Comment by zdcao121** [comment]
Thanks for your reply! I will check it and explore this further.
